[AIRFLOW-3701] Google Cloud Vision Product Search operators

sprzedwojski opened the pull request:
For internal review only so far (since the Cloud Build is failing anyway).

mik-laj commented on airflow/contrib/example_dags/example_gcp_vision.py:
```diff
@@ -0,0 +1,124 @@
+    CloudVisionProductSetGetOperator, CloudVisionProductSetUpdateOperator, \
+    CloudVisionProductSetDeleteOperator, CloudVisionProductCreateOperator, CloudVisionProductGetOperator, \
+    CloudVisionProductUpdateOperator, CloudVisionProductDeleteOperator
+
```
I personally prefer to use a different syntax when there are many classes.
```python
from airflow.contrib.operators.gcp_vision_operator import (
    CloudVisionProductSetCreateOperator, 
    CloudVisionProductSetGetOperator, 
    CloudVisionProductSetUpdateOperator, 
    CloudVisionProductSetDeleteOperator, 
    CloudVisionProductCreateOperator, 
    CloudVisionProductGetOperator, 
    CloudVisionProductUpdateOperator, 
    CloudVisionProductDeleteOperator
)
```
It is cleaner and works better with a git.
This is not a caveat, but a small suggestion.

sprzedwojski replied:
Yeah that's a problem, but I changed it anyway according to the suggestion as it does look much cleaner.

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,257 @@
+
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+from google.cloud.vision_v1 import ProductSearchClient
```
> Imports should be grouped in the following order:
> 
> - Standard library imports.
> - Related third party imports.
> - Local application/library specific imports.
> 
> You should put a blank line between each group of imports.

https://www.python.org/dev/peps/pep-0008/#imports

sprzedwojski replied:
Done

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,257 @@
+            name = ''
+            try:
+                name, product_set_id = self._get_autogenerated_id(response)
+            except Exception as e:
```
This is a very high class. Is there a more suitable exception?

sprzedwojski replied:
I changed it, the exception is more specific

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,257 @@
+            # Product set id was generated by the API
+            name = ''
+            try:
+                name, product_set_id = self._get_autogenerated_id(response)
```
This function is called twice. In both cases it is surrounded by a similar code. Can it be part of the function?

mik-laj commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,603 @@
+
+    If labels are updated, the change will not be reflected in queries until the next index time.
+
+    Note:
```
```suggestion
    .. note:: To locate the `Product` resource, its `name` in the form...........
```
See example: https://raw.githubusercontent.com/apache/airflow/master/docs/howto/initialize-database.rst

sprzedwojski replied:
Done

mik-laj commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,603 @@
+
+    - Returns NOT_FOUND if the Product does not exist.
+    - Returns INVALID_ARGUMENT if display_name is present in update_mask but is missing from the request or
+    longer than 4096 characters.
```
```suggestion
        longer than 4096 characters.
```
See: http://docutils.sourceforge.net/docs/user/rst/quickref.html#bullet-lists

sprzedwojski replied:
Done

mik-laj commented on docs/howto/operator.rst:
```diff
@@ -1652,3 +1652,401 @@ More information
+Makes changes to a :code:`ProductSet` resource. Only :code:`display_name` can be updated
+currently.
+
+Note:
```
As above.

sprzedwojski replied:
Done

mik-laj commented on docs/howto/operator.rst:
```diff
@@ -1652,3 +1652,401 @@ More information
+If labels are updated, the change will not be reflected in queries until the next index
+time.
+
+Note:
```
As above

sprzedwojski replied:
Done

mik-laj commented on docs/howto/operator.rst:
```diff
@@ -1652,3 +1652,401 @@ More information
+Note:
+
+To locate the `Product` resource, its `name` in the form
+`projects/PROJECT_ID/locations/LOC_ID/products/PRODUCT_ID` is necessary.
```
```suggestion
``projects/PROJECT_ID/locations/LOC_ID/products/PRODUCT_ID`` is necessary.
```

sprzedwojski replied:
Done

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductSetCreateOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetCreateOperator
```
Missing `no-index` flag.

sprzedwojski replied:
Done

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductSetGetOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetGetOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductSetUpdateOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetUpdateOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductSetDeleteOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetDeleteOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductCreateOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductCreateOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductGetOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductGetOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductUpdateOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductUpdateOperator
```
Missing `no-index` flag.

mik-laj commented on docs/integration.rst:
```diff
@@ -1511,6 +1511,70 @@ GCPTransferServiceHook
+
+CloudVisionProductDeleteOperator
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductDeleteOperator
```
Missing `no-index` flag.

mik-laj commented:
> airflow/contrib/operators/gcp_vision_operator.py:docstring of airflow.contrib.operators.gcp_vision_operator.CloudVisionProductUpdateOperator:22: WARNING: Bullet list ends without a blank line; unexpected unindent.

While building the documentation, I get an warning message. It should be resolved.

mik-laj commented:
You should add a information about created classes in file ``docs/code.rst``. The previous PRs also lacked this information, but I completed it.

sprzedwojski commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,603 @@
+    :type product_set: dict or google.cloud.vision_v1.types.ProductSet
+        https://googleapis.github.io/google-cloud-python/latest/vision/gapic/v1/types.html#google.cloud.vision_v1.types.ProductSet
+    :param location: (Required) The region where the ProductSet should be created. Valid regions are:
+        us-east1, us-west1, europe-west1, asia-east1
```
The problem is: that information is not in the official documentation...
I extracted it from the error message from the API. I thought it might be better to give this to the user than give nothing.

mik-laj replied:
I think that we should not mention it. I support @potiuk . The difficult art of writing documentation is to write it in such a way that it is always fresh. The documentation should not write about things that are subject to change. Examples: prices, areas of activity.

sprzedwojski replied:
On the other hand we want to be as "user-friendly" as possible. For me (as a user while developing the operators), when testing the Cloud Vision API it was a pain that I couldn't find valid `location`s anywhere in the docs. So I decided to make it easier for the user and include it in the docstring.

I agree that ideally we should be able to link to the Cloud Vision docs, but sadly they don't contain that information. I believe this is better than nothing.

In terms of it being "fresh", as we discussed previously in the case of other operators, it is generally up to the "community" to update the docs when someone notices that it doesn't correspond to the reality anymore.

Also I could add information on when this information was accurate, i.e.:
`Valid regions (as of 2019-02-05) are:`

What do you think? I think it's better than nothing. Also we used to do similar things in the past for other operators, by copying some range of accepted values from the docs, if I recall correctly.

potiuk replied:
Agree. "as of" might be a good idea.

sprzedwojski replied:
Done

sprzedwojski commented on airflow/hooks/base_hook.py:
```diff
@@ -80,7 +80,7 @@ def get_connection(cls, conn_id):
         if conn.host:
             log = LoggingMixin().log
-            log.info("Using connection to: %s", conn.debug_info())
+            log.debug("Using connection to: %s", conn.debug_info())
```
I can revert it. I changed it because when I was running my DAG this was polluting my log, displaying information about parsed DAGs that I wasn't running at the moment.
Depends if we want the logs to be super clean or not.

potiuk replied:
I'd propose it as a very small separate change and see if it will be accepted ;). This way even if it is merged and someone complains, only this change could be reverted easily.

sprzedwojski replied:
Ok, I'll withdraw it from this PR.

sprzedwojski commented on docs/howto/operator.rst:
```diff
@@ -1652,3 +1652,401 @@ More information
+
+- Returns INVALID_ARGUMENT if `display_name` is missing or longer than 4096 characters.
+- Returns INVALID_ARGUMENT if `description` is longer than 4096 characters.
+- Returns INVALID_ARGUMENT if `product_category` is missing or invalid.
```
👍

mik-laj commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,601 @@
+
+    Note:
+
+    To locate the `ProductSet` resource, its `name` in the form
```
Use sphinx's  directive `::note:'.

sprzedwojski replied:
Done

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,254 @@
+            name = response.name
+        except AttributeError as e:
+            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
+        self.log.info('name: ' + name)
```
```suggestion
        self.log.info('name: %s', name)
```
I would also consider the whole form of this message.

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,254 @@
+            if isinstance(e, AlreadyExists):
+                raise e
+            else:
+                self.log.error('The request failed, details below:')
```
You can not expect logs to also contain an exception. In special cases, it will not be true.

sprzedwojski replied:
Done, I changed it.

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,254 @@
+        response = self._handle_request(lambda **kwargs: client.create_product(**kwargs), parent=parent,
+                                        product=product, product_id=product_id, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('Product created:\n%s', response)
```
The whole response should be at a lower level.

sprzedwojski replied:
Done

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,254 @@
+        response = self._handle_request(lambda **kwargs: client.update_product_set(**kwargs),
+                                        product_set=product_set, update_mask=update_mask, retry=retry,
+                                        timeout=timeout, metadata=metadata)
+        self.log.info('ProductSet updated:\n%s', response)
```
The whole response should be at a lower logging level.

sprzedwojski replied:
Done

mik-laj commented:
I finished my review.

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_vision.py:
```diff
@@ -0,0 +1,124 @@
+"""
+import os
+
+from google.api_core.retry import Retry
```
Done

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_vision.py:
```diff
@@ -0,0 +1,124 @@
+GCP_VISION_LOCATION = os.environ.get('GCP_VISION_LOCATION', 'europe-west1')
+# [END howto_operator_vision_args_common]
+
+product_set = ProductSet(display_name='My Product Set 1')
```
Done

sprzedwojski commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,257 @@
+        return product_set
+
+    @staticmethod
+    def _determine_product_name(product, product_id, location, project_id):
```
Done, refactored to a single `_determine_name()` method.

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,254 @@
+        try:
+            return fun(**kwargs)
+        except GoogleAPICallError as e:
+            if isinstance(e, AlreadyExists):
```
It looks good for me. @potiuk, @antonimaciej  what do you think?

potiuk replied:
It's ok to use several except clauses, but I think the problem is that we are raising either Google API "AlreadyExist" or AirflowException from the same method. Another approach could be to have different exception deriving from AirflowException: for example ResourceAlreadyExist(AirflowException). This one should be thrown instead of google API AlreadyExist raised. This way we always have AirflowException from this method and it better separates concerns.

I also do not see much value in self.log.error() message as it is now :).

sprzedwojski replied:
In the original version (with `isinstance`) I think it is clearer that `AlreadyExists` inherits from `GoogleAPICallError`.

potiuk replied:
We discussed before that wrapping an exception with AirflowException helps with getting the right stack trace (this error is not always ignored - sometimes it bubbles up). I think we agreed before that we wrap the original low-level exceptions with AirflowException in our hooks (rather than let it happen outside of "execute" method in airflow core) to get the right stack trace - this is what @antonimaciej was working on before. I think we should be consistent :)

I am not insisting on having this special exception but I am agains bubbling up the AlreadyExist error to the operator. Are we 100% sure "AlreadyExist" will always be handled separately and ignored in all current and future methods calling _handle_request?

I actually looked a bit more and I found related problem (another comment follows shortly).

potiuk replied:
At the very least we should at least mention in the docs that AlreadyExists error might be returned and should be handled (in the docstrings of all methods using _handle_request that might return it).

sprzedwojski replied:
> Are we 100% sure "AlreadyExist" will always be handled separately and ignored in all current and future methods calling _handle_request?
That's why I like the verbosity of Java method signatures which list all checked exception that it can throw ;)

I looked at our previous operators from Q4 2018 and there's a lot of try/except clauses checking for exceptions different than `AirflowException`.

That said, I could handle the `AlreadyExists` error in the hook, however to me such business logic belongs in the operator.
In my opinion the hook should be "stupid", being the service layer only, and the operator should interpret its responses and behave accordingly if necessary. And that's what we've generally done in the previous operators.

**TL;DR: I see nothing bad in handling some exceptions in the operators.**

mik-laj commented on docs/code.rst:
```diff
@@ -205,6 +205,14 @@ Operators
 .. autoclass:: airflow.contrib.operators.gcp_sql_operator.CloudSqlInstanceImportOperator
 .. autoclass:: airflow.contrib.operators.gcp_sql_operator.CloudSqlInstancePatchOperator
 .. autoclass:: airflow.contrib.operators.gcp_sql_operator.CloudSqlQueryOperator
+.. autoclass:: airflow.contrib.operators.gcp_vision_operator.CloudVisionProductSetCreateOperator
```
Sort alphabetically by class name and then package.

sprzedwojski replied:
Done

potiuk commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,599 @@
+        except AlreadyExists:
+            self.log.info('Product set with id %s already exists. Aborting the create operation.',
+                          self.product_set_id)
+            return False
```
I do not think we should return False here - we should return product_set_id to keep idempotency

sprzedwojski replied:
Done

potiuk commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,599 @@
+                                                 timeout=self.timeout,
+                                                 metadata=self.metadata)
+        except AlreadyExists:
+            self.log.info('Product set with id %s already exists. Aborting the create operation.',
```
Aborting -> Finishing. It's not an error.

sprzedwojski replied:
Done, changed it to `Exiting from`.

potiuk commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,599 @@
+        except AlreadyExists:
+            self.log.info('Product with id %s already exists. Aborting the create operation.',
+                          self.product_id)
+            return False
```
Same here. I believe we should return product_id here.

potiuk commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,599 @@
+                                             timeout=self.timeout,
+                                             metadata=self.metadata)
+        except AlreadyExists:
+            self.log.info('Product with id %s already exists. Aborting the create operation.',
```
Aborting -> Finishing.

sprzedwojski replied:
Done, changed it to `Exiting from`.

sprzedwojski commented on airflow/contrib/example_dags/example_gcp_vision.py:
```diff
@@ -0,0 +1,124 @@
+    # [START howto_operator_vision_product_set_create]
+    product_set_create = CloudVisionProductSetCreateOperator(
+        location=GCP_VISION_LOCATION,
+        product_set=product_set,
```
Done, I've duplicated all operators in the example DAG and each one has 2 versions:
- with autogenerated id
- with explicitly provided id

I've also added those snippets to the `operator.rst` docs.

mik-laj commented on tests/contrib/operators/test_gcp_vision_operator.py:
```diff
@@ -0,0 +1,181 @@
+from google.api_core.exceptions import AlreadyExists
+from google.cloud.vision_v1.proto.product_search_service_pb2 import ProductSet, Product
+
+from airflow.contrib.operators.gcp_vision_operator import CloudVisionProductSetCreateOperator, \
```
Here you can change the style to the one with brackets.

sprzedwojski replied:
Done

mik-laj commented on airflow/contrib/operators/gcp_vision_operator.py:
```diff
@@ -0,0 +1,599 @@
+    :param retry: (Optional) A retry object used to retry requests. If `None` is
+        specified, requests will not be retried.
+    :type retry: google.api_core.retry.Retry
+        https://googleapis.github.io/google-cloud-python/latest/core/retry.html#google.api_core.retry.Retry
```
Nie należy dodawać linków, ale włączyć wewnętrzne linkowanie.Tutaj przykład: https://github.com/apache/airflow/pull/4655

sprzedwojski replied:
Sądzę, że nie jest to póki co konwencja w Airflow, więc na razie zostawię tak, jak jest.
Jak wejdą Twoje zmiany to można to wyedytować.

sprzedwojski commented:
Thanks for the review, guys. I've addressed all the comments. Would be great if you could take a look again and LGTM it.
Once you do that I'll squash the commits and create a PR to Apache.

mik-laj commented on docs/integration.rst:
```diff
@@ -701,6 +701,77 @@ Transfer Service
+Cloud Vision
+''''''''''''
+
+Cloud Vision Product Search Operators
```
Please rebase.

sprzedwojski replied:
What do you mean? I've rebased on top of `apache/master` and there were no conflicts.

sprzedwojski replied:
I've updated the `integration.rst` to the new format (only linking to `code.rst`).

sprzedwojski commented:
Thanks for all the feedback. I'm closing this PR and I've opened up the [PR to Apache](https://github.com/apache/airflow/pull/4665).

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+from airflow import AirflowException
+from airflow.contrib.hooks.gcp_api_base_hook import GoogleCloudBaseHook
+
+PARENT_TEMPLATE = 'projects/{}/locations/{}'
```
I think, I think you should use 
https://googleapis.github.io/google-cloud-python/latest/vision/gapic/v1/api.html#google.cloud.vision_v1.ProductSearchClient.location_path
https://googleapis.github.io/google-cloud-python/latest/vision/gapic/v1/api.html#google.cloud.vision_v1.ProductSearchClient.product_path
https://googleapis.github.io/google-cloud-python/latest/vision/gapic/v1/api.html#google.cloud.vision_v1.ProductSearchClient.product_set_path

sprzedwojski replied:
Thanks, good call. I've implemented it in https://github.com/apache/airflow/pull/4665

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+
+    @staticmethod
+    def _determine_name(entity, entity_id, location, project_id):
+        is_product = type(entity) == Product
```
This is not the right solution. You should take logic to a separate class and parameterize it, because you are now creating a code that is very hard to develop and maintain.

mik-laj replied:
This function does not do one thing, but it does two at the same time, but the problem is just similar.

mik-laj replied:
The logic is too complicated in my opinion. I have a big problem to understand all the cases that are included in this code.  Look at this: https://gitlab.polidea.com/snippets/18

mik-laj replied:
I respect that you may have a different opinion and invite you to a discussion.

mik-laj replied:
Look at:  https://gitlab.polidea.com/snippets/19

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+            if not explicit_name:
+                entity.name = template_name
+            elif explicit_name != template_name:
+                raise AirflowException("The {} name provided in the object ({}) is different than the"
```
If you have a lot of arguments, use the keywords arguments to increase readability
```python
                raise AirflowException("The {entity_label} name provided in the object ({explicit_name}) is different than the"
                                       " name created from the input parameters ({template_name}). Please either: "
                                       "1) Remove the {entity_label} name, 2) Remove the location and "
                                       "{entity_id_label} parameters, 3) Unify the {entity_label} name and input "
                                       "parameters.".format(entity_label=entity_label, entity_id_label=entity_id_label, template_name=template_name, explicit_name=explicit_name))
```
However, if you really do not want to, you can use explicit numbering. 
```python
>>> "{1}{0}{0}".format("A", "B")
'BAA'
````

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+
+from google.api_core.exceptions import AlreadyExists, GoogleAPICallError, RetryError
+from google.cloud.vision_v1 import ProductSearchClient
+from google.cloud.vision_v1.proto.product_search_service_pb2 import Product
```
```python
from google.cloud.vision_v1.types import Product
````

sprzedwojski replied:
This doesn't work https://drive.google.com/open?id=1kGrF1nhhq90-g-wLYhcFnoAqDV6p7-y6

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+                                       "{} parameters, 3) Unify the {} name and input "
+                                       "parameters.".format(entity_label, explicit_name, template_name,
+                                                            entity_label, entity_id_label, entity_label))
+        return entity
```
It's required?

sprzedwojski replied:
Yes, because when invoking the function:
`product = self._determine_name(product, product_id, location, project_id)`

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+            raise AirflowException(e)
+
+    @staticmethod
+    def _get_autogenerated_id(response):
```
If I understand it correctly, you can shorten it to
```python
name = response.get('name')
if '/' not in name:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
return name.rsplit('/', 1)[1]
```

sprzedwojski replied:
After having added all the necessary validations the difference is minimal, 8 lines versus 10:
```
try:
            name = response.name
        except AttributeError as e:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
        product_set_id = name.rsplit('/', 1)[1]
        if '/' not in name or not product_set_id:
            raise AirflowException('Unable to get id from name... [{}]'.format(name))
        return product_set_id
```

But I'll change it.

sprzedwojski replied:
Ok, after adding some more tests I shortened it to 7 lines ;)
```
@staticmethod
    def _get_autogenerated_id(response):
        try:
            name = response.name
        except AttributeError as e:
            raise AirflowException('Unable to get name from response... [{}]\n{}'.format(response, e))
        if '/' not in name:
            raise AirflowException('Unable to get id from name... [{}]'.format(name))
        return name.rsplit('/', 1)[1]
```

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+        else:
+            template_name = entity_name_template.format(project_id, location, entity_id)
+            if not explicit_name:
+                entity.name = template_name
```
I think, you should avoid modify objects from user.  He may not know about it and want to use the same object in the loop.
https://softwareengineering.stackexchange.com/questions/245767/is-modifying-an-incoming-parameter-an-antipattern

sprzedwojski replied:
I can agree on that. I effectively mixed two things: I do return the full object (see https://github.com/PolideaInternal/airflow/pull/51/files/e6868c2bb5d0154891dd0f8b0f1b08292c648b98#r257625543), which is a good approach, but I also changed the object passed in by the user.

So I added a deepcopy at the beginning of this function, and this way I operate on a copy of the object, thus not modifying the user's object, and at the end of the function I return the modified object to the user.

`entity = deepcopy(entity)`

mik-laj commented on airflow/contrib/hooks/gcp_vision_hook.py:
```diff
@@ -0,0 +1,245 @@
+from google.api_core.exceptions import AlreadyExists, GoogleAPICallError, RetryError
+from google.cloud.vision_v1 import ProductSearchClient
+from google.cloud.vision_v1.proto.product_search_service_pb2 import Product
+from google.protobuf.json_format import MessageToJson
```
MessageToDict
